### PR TITLE
Merkle Mountain Range 

### PR DESCRIPTION
**Pull request by oshorefueled**
This PR adds the MMR proof implementation

### CODE REVIEW

File: mmr/mmr.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	"sort"
+)
+
+type MMR struct {
+	merge func(left, right interface{}) interface{}
+}
+
+func (m *MMR) calculatePeakRoot(leaves []leaf, peakPos uint64, proofs *Iterator) (interface{}, error) {
+	if len(leaves) == 0 {
+		// TODO: clarify on how debug_assert! works
```
Comment (seunlanlege) [inline mmr/mmr.go:14]: remove todo

File: mmr/types.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+type leafWithHash struct {
+	pos    uint64
+	hash   interface{}
+	height uint32
+}
+
+type peak struct {
+	height uint32
+	pos    uint64
+}
```
Comment (seunlanlege) [inline mmr/types.go:17]: Camelcase names?

**Review by seunlanlege — CHANGES_REQUESTED** [review]

File: mmr/types.go (hunk 2 above)
Comment (oshorefueled) [reply mmr/types.go:17]: yes, camelcase names is the general practise in go

File: mmr/mmr.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
+package mmr
+
+import (
+	"fmt"
+	"sort"
+)
+
+type MMR struct {
+	merge func(left, right interface{}) interface{}
```
Comment (seunlanlege) [inline mmr/mmr.go:9]: Let's use the merge interface mehdi created.

**Review by seunlanlege — CHANGES_REQUESTED** [review]

**Pull request merged**